### PR TITLE
Restore concurrency during publishing

### PR DESCRIPTION
**Pull request by djoyner**
Addresses post-merge feedback from @yanfali on #119.

This PR restores the ability of realm-level session handlers to publish concurrently.  Includes commit from #117.

### CODE REVIEW

File: broker.go [hunk 1]
```diff
@@ -56,20 +62,24 @@ func (br *defaultBroker) Publish(pub Sender, msg *Publish) {
...
+	br.lock.Unlock()
 
 	sub.Send(&Subscribed{Request: msg.Request, Subscription: id})
 }
 
 func (br *defaultBroker) Unsubscribe(sub Sender, msg *Unsubscribe) {
+	br.lock.Lock()
 	topic, ok := br.subscriptions[msg.Subscription]
 	if !ok {
+		br.lock.Unlock()
```
Comment (yanfali) [inline broker.go:82]: shouldn't you just use defer?

Comment (yanfali) [inline broker.go:82]: never mind, I see you're trying to not lock the Send. You might want to use a scoping block to make defer work just for this scope.

**Comment by yanfali** [comment]
Wow, very different patch now. How are you testing this?

**Comment by djoyner** [comment]
@yanfali Yes, it is quite a bit different now. In order to allow the realm-level session handlers to execute concurrently the broker and dealer had to be made safe.  I would have preferred a goroutine-based approach to handling concurrency as in the original PR.  But that would require some additions to the `Broker` and `Dealer` interfaces (e.g. they could run internal goroutines but then would need a `Close` method for clean shutdown).  Since those are existing interfaces this would be a breaking change and I felt it best to avoid that.  So I opted for internal locking in the broker/dealer.

You commented earlier about locks around `Send`.  The only `Send` that is covered by a lock is in the broker's `Publish` method as it is ranging over `routes`.  I contemplated building a separate slice of subscribers under lock protection and then iterating to send outside the lock.  But that would create garbage and seemed unnecessary since the read-lock allows `Publish` to execute concurrently and this is the common case.  The write lock is only used during `Subscribe`/`Unsubscribe` and there I was careful not to `Send` while holding the lock.

As far as testing goes, I took a three-pronged approach:
1. `go vet *.go` (not testing, but still helpful)
2. `go test -race *.go`
3. My company uses `turnpike` client and server in a large back-end implementation.  We have several integration tests that cover publish, subscribe, register, call, etc.

**Comment by yanfali** [comment]
Very cool. I did my initial testing with race, but I actually ran my deployment with -race too and found a bunch more race conditions, especially in localPeer. Apparently it connects before the realm is fully set up. You might want to try running one of your test servers using -race, you may be surprised what you find.

**Comment by djoyner** [comment]
@yanfali Interesting. At one point we tried our test deploys with -race but ran into so many issues with turnpike it led to #118 and this whole effort.  I have not gone back to that effort but probably should. 

It sounds like you have some good fixes in hand. Why not bring a PR for some of these yourself? I'm sure people would be appreciative.

**Comment by yanfali** [comment]
I have them in a branch, on a public repo, I'm concerned they won't fit stylistically with the way you are solving the same issues, but I'll try and port over the one with localPeer and realm, they could definitely use more eyes on them to see if the approach is palatable to a wider audience.

**Comment by yanfali** [comment]
https://github.com/awakenetworks/turnpike/commit/87b42a9603502ae827f383a627018388ce5c9db0 is the specific patch. Code not withstanding, the reasoning behind the patch is described in the comment. It would be good for a second person to confirm that conclusion. @djoyner

**Pull request merged**